### PR TITLE
ZCS-4142:Back-end javascript resource consolidation and filtering for…

### PR DESCRIPTION
**Pull request by sneha-patil-synacor**
Back-end javascript resource consolidation and filtering for next gen zimlets

- Using closure compiler library for consolidation of next gen zimlets 
- next gen zimlets manifest file will include an attribute called 'zimbraXZimletCompatibleSemVer' - which will have a SemVer Expression
- To request the consolidated javascript for only next gen compatible zimlets, the query parameter to existing URL will be specified e.g. zimbraXZimletCompatibleWith=4.1.2
- Semantic version check will be done using jsemver java library during zimlet filtering process
- Only zimbraX zimlets script compilation will be done using closure-compiler library, older zimlets will continue using yui compressor

### CODE REVIEW

File: store/conf/attrs/zimbra-attrs.xml [hunk 1]
```diff
@@ -9640,4 +9640,7 @@ TODO: delete them permanently from here
   <globalConfigValue>1518163473</globalConfigValue>
 </attr>
 
+<attr id="3024" name="zimbraXZimletCompatibleSemVer" type="string" max="256" cardinality="single" optionalIn="zimletEntry" since="8.8.8">
```
Comment (rupalid) [inline store/conf/attrs/zimbra-attrs.xml]: change it to 8.8.9

File: store/ivy.xml [hunk 2]
```diff
@@ -104,5 +104,7 @@
   <dependency org="zimbra" name="zm-native" rev="latest.integration"/>
   <dependency org="org.xmlunit" name="xmlunit-core" rev="2.3.0"/>
   <dependency org="ant-contrib" name="ant-contrib" rev="1.0b3" />
+  <dependency org="com.google.javascript" name="closure-compiler" rev="v20180204"/>
```
Comment (rupalid) [inline store/ivy.xml]: Create tickets to include them as part of zm-build., I assume they go into /opt/zimbra/lib/jars

File: store/src/java/com/zimbra/cs/zimlet/ZimletFilter.java [hunk 3]
```diff
@@ -242,6 +243,24 @@ public void doFilter(ServletRequest request, ServletResponse response,
...
+                .valueOf(zimbraXZimletCompatibleWith);
+            boolean result = v.satisfies(zimlet.getZimbraXCompatibleSemVer());
+            if (result) {
+                zimletList.add(zimlet);
+            }
+        } else if (zimbraXZimletCompatibleWith == null
+            && zimlet.getZimbraXCompatibleSemVer() == null) {
+            // request is not for ZimbraX zimlets and this is not zimbraX zimlet
+            zimletList.add(zimlet);
+        }
```
Comment (rupalid) [inline store/src/java/com/zimbra/cs/zimlet/ZimletFilter.java]: Do we need to handle the case when only one of them is null

File: store/conf/attrs/zimbra-attrs.xml (hunk 1 above)
Comment (sneha-patil-synacor) [reply store/conf/attrs/zimbra-attrs.xml]: Fixed

File: store/ivy.xml (hunk 2 above)
Comment (sneha-patil-synacor) [reply store/ivy.xml]: Created ZCS-4808

**Pull request merged**